Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -479,103 +498,148 @@ void CGContextAddRects(CGContextRef context, const CGRect* rects, unsigned count
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathAddArcToPoint(context->Path(), &(context->CurrentGState().transform), x1, y1, x2, y2, radius);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddEllipseInRect(CGContextRef context, CGRect rect) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathAddEllipseInRect(context->Path(), &(context->CurrentGState().transform), rect);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddPath(CGContextRef context, CGPathRef path) {
     NOISY_RETURN_IF_NULL(context);
-    // The Apple SDK docs imply that passing a NULL path is valid.
-    // So avoid calling into the backing if NULL.
-    if (path) {
-        UNIMPLEMENTED();
+    if (!path) {
+        return;
     }
+
+    if (!context->HasPath()) {
+        // If we don't curerntly have a path, take this one in as our own.
+        woc::unique_cf<CGMutablePathRef> copiedPath{ CGPathCreateMutableCopy(path) };
+        context->SetPath(copiedPath.get());
+        return;
+    }
+
+    CGPathAddPath(context->Path(), &(context->CurrentGState().transform), path);
 }
 
 /**
  @Status Stub
 */
 void CGContextReplacePathWithStrokedPath(CGContextRef context) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    if (!context->HasPath()) {
+        return;
+    }
+
+    auto& state = context->CurrentGState();
+
+// TODO GH#xxxx When CGPathCreateCopyByStrokingPath is no longer stubbed, remove the diagnostic suppression.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+
+    woc::unique_cf<CGPathRef> newPath{ CGPathCreateCopyByStrokingPath(context->Path(),
+                                                                      nullptr, // The points in the path are already transformed; do not transform again!
+                                                                      state.lineWidth,
+                                                                      (CGLineCap)state.strokeProperties.startCap,
+                                                                      (CGLineJoin)state.strokeProperties.lineJoin,
+                                                                      state.strokeProperties.miterLimit) };
+
+#pragma clang diagnostic pop
+
+    woc::unique_cf<CGMutablePathRef> newMutablePath{ CGPathCreateMutableCopy(newPath.get()) };
+    context->SetPath(newMutablePath.get());
 }
 
 /**
  @Status Interoperable
 */
 bool CGContextIsPathEmpty(CGContextRef context) {
     NOISY_RETURN_IF_NULL(context, StubReturn());
-    UNIMPLEMENTED();
-    return StubReturn();
+
+    return !context->HasPath() || CGPathIsEmpty(context->Path());
 }
 
 /**
  @Status Interoperable
 */
 CGRect CGContextGetPathBoundingBox(CGContextRef context) {
-    NOISY_RETURN_IF_NULL(context, StubReturn());
-    UNIMPLEMENTED();
-    return StubReturn();
+    NOISY_RETURN_IF_NULL(context, CGRectNull);
+
+    if (!context->HasPath()) {
+        return CGRectNull;
+    }
```

Comment:
To be the devil: RETURN_RESULT_IF(!context->HasPath(),CGRectNull);

You can use RETURN_IF(), etc for the non noisy conditional checks.